### PR TITLE
PyPy compatibility

### PR DESCRIPTION
**Pull request by tilgovi**
This changeset addresses #49 by fixing some small issues to enable PyPy3.5 compatibility.

Some changes are simple and make sense, such as changes to address the differences in garbage collection or the absence of global reference debugging in PyPy.

The one change that is troubling is the change to call the clear functions directly from the deallocation functions in the `Handle` and `Loop` implementations. It seems that under PyPy, the `tp_clear` member can be `NULL` in the basetype inheritance tests. I don't know why this is yet nor whether it may be caused by a bug in cpyext. However, looking at the Python 3.6 source tree, I see explicit calls to clear implementations in deallocators rather than calls delegated through the type object. I would just want to know that this doesn't break any garbage collection behavior for subclasses.

Note that there's a nasty `before_install` script for Travis because Travis hasn't updated its PyPy version yet.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -10,6 +10,10 @@ matrix:
           env: TOX_ENV=py34
         - python: 3.5
           env: TOX_ENV=py35
+        - python: pypy3
```
Comment (saghul) [inline .travis.yml]: can you also add pypy 2?

File: .travis.yml [hunk 2]
```diff
@@ -10,6 +10,10 @@ matrix:
...
+          env: TOX_ENV=pypy3
+before_install:
+    - "if [[ $TRAVIS_PYTHON_VERSION == 'pypy3' ]]; then deactivate && cd $HOME && wget https://bitbucket.org/squeaky/portable-pypy/downloads/pypy3.5-5.7-beta-linux_x86_64-portable.tar.bz2 && tar -jxvf pypy3.5-5.7-beta-linux_x86_64-portable.tar.bz2 && cd pypy3.5-5.7-beta-linux_x86_64-portable && echo 'Setting up aliases...' && export PATH=$HOME/pypy3.5-5.7-beta-linux_x86_64-portable/bin/:$PATH && ln -s ~/pypy3.5-5.7-beta-linux_x86_64-portable/bin/pypy3.5 bin/python && echo 'Setting up pip...' && bin/pypy -m ensurepip && ln -s ~/pypy3.5-5.7-beta-linux_x86_64-portable/bin/pip3.5 bin/pip && cd $TRAVIS_BUILD_DIR; fi"
```
Comment (saghul) [inline .travis.yml]: can we turn this into a shell script in some .travis directory perhaps? then we'd just invoke it here.

**Comment by saghul** [comment]
I see there a weird failing test, which seems unrelated. I'll merge this shortly, thanks a lot for working on it! ❤️

**Comment by tilgovi** [comment]
Okay. I've done as you asked.

Recent PyPy uses the same code for the sys module helpers on 2.7 and 3.5 and so the stderr printing can sometimes pass unicode objects to `StdBufferIO.write()`, so that required a small change.

There is only one remaining issue I see. It may help if you explain the purpose of the weakref test (as opposed to the gc test).

On PyPy, the deallocate function is called only once, even if we try to resurrect the object. From my reading of the PyPy codebase, there's no check after the deallocation function is called to see if it was resurrected and the `pyuv__handle_dealloc_close_cb` is never called.

Give me a chance to look into this a little bit more before you merge. I want to see that libuv isn't getting confused and failing to close pending handles.

This problem shouldn't cause leaked file descriptors, because those are closed immediately by `uv_close` (at least on *nix), but if it breaks the close pending linked list then other callbacks might get dropped.

**Comment by tilgovi** [comment]
@saghul thanks for your patience. I think I've got it all sorted out now.

After spending a lot of time digging in the PyPy and CPython GC and C extension code, and stepping through the tests under gdb, I decided that removing the virtual dispatch of `Py_TYPE(self)->tp_clear` was a dangerous idea that might result in leaks. Instead, we can account for the lack of this member on PyPy subtypes by traversing the base links until we find one.

However, I discovered that PyPy will invalidate weakrefs whether or not we resurrect an object in its deallocator. Therefore, I did my best to convert the weakref test into something that attempts to more directly test that the resurrection works as intended. Unfortunately, I could find no way to actually assert that the objects had not been freed except to assume that touching them would segfault otherwise.

I also took the time to rebase the commits into clean pieces and change their messages to match the style of the project.

Let me know if there's anything else I can do!

**Comment by saghul** [comment]
This is fantastic work, thank you so much! I'll take a closer look and merge tonight, but all tests pass so we might now be able to close the oldest issue in the repo! 🎉

**Pull request merged**